### ISSUE TITLE
Train Emformer RNN-T using provided recipe cannot converge 

### ISSUE THREAD

**Issue by thanhtvt** [pytorch/audio#2613]
### 🐛 Describe the bug

Thank you for the amazing implementation of Emformer and the [training recipe](https://github.com/pytorch/audio/tree/main/examples/asr/emformer_rnnt) of it. 
But when I train from scratch using this recipe, the model cannot converge and stop improving after only about 6-7 epochs (tensorboard image below).

![tensorboard](https://user-images.githubusercontent.com/62584583/184278140-a8c38214-748f-421b-90ca-948601ff63cb.png)

I trained on 100-hour training set of Librispeech. I also have re-created `global_stats.json` file to fit to only 100-hour data using this [script](https://github.com/pytorch/audio/blob/main/examples/asr/emformer_rnnt/global_stats.py) (I have commented out other LibriSpeech sets of course)

Is there something I miss or something? Can you please check it out for me? Thank you very much!

### Versions

```
PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.13 (default, Mar 28 2022, 11:38:47)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-52-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: A100-SXM4-40GB
GPU 1: A100-SXM4-40GB
GPU 2: A100-SXM4-40GB
GPU 3: A100-SXM4-40GB
GPU 4: A100-SXM4-40GB
GPU 5: A100-SXM4-40GB
GPU 6: A100-SXM4-40GB
GPU 7: A100-SXM4-40GB

Nvidia driver version: 450.80.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.1
[pip3] pytorch-lightning==1.7.1
[pip3] torch==1.12.1
[pip3] torch-tb-profiler==0.4.0
[pip3] torchaudio==0.12.1
[pip3] torchmetrics==0.9.3
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.6.0              hecad31d_10    conda-forge
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py38h95df7f1_0    conda-forge
[conda] mkl_fft                   1.3.1            py38h8666266_1    conda-forge
[conda] mkl_random                1.2.2            py38h1abd341_0    conda-forge
[conda] numpy                     1.23.1           py38h6c91a56_0  
[conda] numpy-base                1.23.1           py38ha15fc14_0  
[conda] pytorch                   1.12.1          py3.8_cuda11.6_cudnn8.3.2_0    pytorch
[conda] pytorch-lightning         1.7.1                    pypi_0    pypi
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch-tb-profiler         0.4.0                    pypi_0    pypi
[conda] torchaudio                0.12.1               py38_cu116    pytorch
[conda] torchmetrics              0.9.3                    pypi_0    pypi
```

**Comment by hwangjeff** [comment]
Hi @thanhtvt, thanks for posting — it's great to hear that you're finding Emformer and the training recipe to be useful. We haven't tried training the model on just the 100-hour LibriSpeech training set. One thought, however, is that the model capacity may be too large relative to the amount of data you're using, which may be causing the overfitting you're observing. Perhaps you can try reducing the size of the model, e.g. reducing the number of Emformer layers, or tune the training configuration. Please keep us updated on how it goes!

**Comment by thanhtvt** [comment]
Hi @hwangjeff , thanks for replying - I have tried to reduced size of the model like you said by modifying parameters of `emformer_rnnt_model` function. Details are below

```python
    self.model = emformer_rnnt_model(
        input_dim=80,
        encoding_dim=512,
        num_symbols=501,
        segment_length=16,
        right_context_length=4,
        time_reduction_input_dim=128,
        time_reduction_stride=4,
        transformer_num_heads=8,
        transformer_ffn_dim=1024,
        transformer_num_layers=16,
        transformer_dropout=0.1,
        transformer_activation="gelu",
        transformer_left_context_length=30,
        transformer_max_memory_size=0,
        transformer_weight_init_scale_strategy="depthwise",
        transformer_tanh_on_mem=True,
        symbol_embedding_dim=512,
        num_lstm_layers=1,
        lstm_layer_norm=True,
        lstm_layer_norm_epsilon=1e-3,
        lstm_dropout=0.3,
    )
```
I have also re-trained sp_model by using [this script](https://github.com/pytorch/audio/blob/main/examples/asr/emformer_rnnt/librispeech/train_spm.py), whose `vocab_size` is 500.

This time, the model is about 36M params and the training is somehow better. However, the validation stopped decreasing after about 37 epochs. I trained it overnight and used checkpoint at 102nd epoch to evaluate. The result was about 18.86% WER on test-clean. This is much lower than I expected (I want it about 8-9% WER). Do you have any suggestion on that? Thank you very much!

Here is my tensorboard
![image](https://user-images.githubusercontent.com/62584583/185029880-746a4f3f-9c09-41c3-91eb-5339193a4a56.png)

**Comment by thanhtvt** [comment]
so... any updates? @hwangjeff

**Comment by hwangjeff** [comment]
You can continue tuning the architecture and training parameters — I don't have any recommendations that are more specific than that. Also note that the relationship between validation loss and validation WER isn't direct, i.e. validation loss might increase while validation WER decreases, so it'd be worth keeping an eye on how the validation WER evolves over the course of training.
